modules/git/batch_reader.go

```diff
@@ -44,10 +35,28 @@ func EnsureValidGitRepository(ctx context.Context, repoPath string) error {
 	return nil
 }
 
+func returnClosedReaderWriters(err error) (WriteCloserError, *bufio.Reader, func()) {
+	wr := &writeCloserError{}
+	rd := &readCloserError{}
+
+	_ = wr.CloseWithError(err)
+	_ = rd.CloseWithError(err)
+
+	return wr, bufio.NewReader(rd), func() {}
+}
+
 // CatFileBatchCheck opens git cat-file --batch-check in the provided repo and returns a stdin pipe, a stdout reader and cancel function
 func CatFileBatchCheck(ctx context.Context, repoPath string) (WriteCloserError, *bufio.Reader, func()) {
-	batchStdinReader, batchStdinWriter := io.Pipe()
-	batchStdoutReader, batchStdoutWriter := io.Pipe()
+	batchStdinReader, batchStdinWriter, err := os.Pipe()
+	if err != nil {
+		log.Critical("Unable to open pipe to write to: %v", err)
+		return returnClosedReaderWriters(err)
+	}
+	batchStdoutReader, batchStdoutWriter, err := os.Pipe()
```

Comment:
We might want to check if this is CPU/Memory-efficient, the current implementation use buffered I/O(for good reason, this can and will be called on bigger files).

Reply:
`os.Pipe` is the correct and proper way. According to Golang source, if the `Stdout` provided is not a `os.File(Pipe)`, then there will be a new internal `os.Pipe` be used in the end.

So if you can provide the os.Pipe to Stdout, it will be used directly.


```go
func (c *Cmd) writerDescriptor(w io.Writer) (f *os.File, err error) {
	if w == nil {
		f, err = os.OpenFile(os.DevNull, os.O_WRONLY, 0)
		if err != nil {
			return
		}
		c.closeAfterStart = append(c.closeAfterStart, f)
		return
	}

	if f, ok := w.(*os.File); ok {
		return f, nil
	}

	pr, pw, err := os.Pipe()
	if err != nil {
		return
	}

	c.closeAfterStart = append(c.closeAfterStart, pw)
	c.closeAfterWait = append(c.closeAfterWait, pr)
	c.goroutine = append(c.goroutine, func() error {
		_, err := io.Copy(w, pr)
		pr.Close() // in case io.Copy stopped due to write error
		return err
	})
	return pw, nil
}
```